### ISSUE TITLE
Interesting error in compute_newfilters

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#87]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/87) by Charles James Titus (Bitbucket: [cjtitus](https://bitbucket.org/cjtitus), GitHub: [cjtitus](https://github.com/cjtitus)).**

----------------------------------------

Due to the large amount of data I am analyzing, it is of high importance to me that MASS be robust enough to leave unattended while ~TBs of data analyze themselves overnight.

In one dataset I am trying to analyze, there is a channel which contains only 26 good pulses. There were 10,000 pulses altogether, so this channel is clearly borked. However, it is never automatically set bad. If it had fewer than 10 good pulses, compute_filters would set it bad. However I will argue that this number is too low, or the other checks are inadequate.

In the ds.compute_newfilters code, up to 1000 good pulses are selected. After computing various things, the arrival time is digitized. 

```
#!python
ATime = ATime[use]
NBINS = 9
bins = np.digitize(ATime, np.linspace(ATime.min(), ATime.max(), NBINS+1))-1
```

However, for small numbers of pulses, it is quite possible that the bins are not all populated. In my example with 26 good pulses, of which only 22 were selected for use, 

```
#!python
print bins
[2 7 1 0 3 2 3 4 3 6 0 6 4 0 8 1 6 4 6 9 4 8]
```

Eagle-eyed observers will notice that "5" is missing. No pulses had an arrival time that falls into bin 5.

This creates a problem, because later on we can have

```
#!python
ymed = [np.median(y[bins==i]) for i in range(NBINS)]
print ymed
[0.044487491, 0.047441784, 0.05048272, 0.052494507, 0.049994569, nan, 0.071980178, 0.090351045, 0.099023506]
```

Oh dear. Notice the nan in there? That's not going to play nicely with 
```
#!python
fit = np.polyfit(xmed, ymed, DEGREE)
```

And you will get a cryptic error 
```
#!python
ValueError: On entry to DGELSD parameter number 4 had an illegal value
```

It would be great if this could be more robust. 
I think possible solutions are:

* Surround polyfit with a try/catch
* Increase required number of good pulses from 10 to 100
* Check inputs to polyfit
* In channel_group, surround the call to ds.compute_newfilter with a try/catch

Obviously, it's easiest to just chuck in a try/catch, but I don't like the way that it makes other errors harder to detect/debug. Increasing required number of good pulses is probably a good idea in any event because here is an example where it is too low.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Hi Jamie. Message received, and I really appreciate your diligent investigation of the problem. I have previously seen that cryptic error and have rarely had any clue what caused it. Unless I have overlooked it, you did not mention whether this problem caused a hard crash of iPython, but I think that it has done so for me in the past. Could you clarify that point?

Okay, so you've discovered that a detector with 22 good pulses is basically unable to make--reliably--a newfilter, because of the NaNs.  You propose two separate fixes (four, actually, but I am ignoring the two that add try/catch):

1. I should make the newfilter-creation step more robust. For example, the (xmed, ymed) vectors could both be handled so that empty bins aren't used in the polyfit call. 
1. We could also raise the required number of good pulses.

I am definitely on board with the first. But how in the world do I know how high to raise the required number? I could arbitrarily choose 50 or 100, but there's no telling whether that's enough for the next problem.

Given that Happy Hour is starting any minute, can I suggest that you impose your own limit for now and proceed from there? I'll get to this next week.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

@oneilg might have something to say about this issue. Meanwhile, I have developed a test and a fix. If anyone wants to raise the mandatory # of good pulses, let's discuss. I'm going to fix by the first of the 2 methods that I mentioned, above.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

When building new filters, do not use empty arrival-time bins.

A linear (or other low-order polynomial) model is built using binned
data, where arrival times are grouped into 9 bins. But as Jamie 
discovered, this procedure horks whenever there are few enough pulses
to leave a bin empty. With this commit, empty bins are ignored (or
even near-empty bins). Fixes #87.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I’m glad you’ve fixed this. I’m hoping we can shortly move to a system where we just use the same filter on every dataset (for the same channel), rather than calculate a new filter every time we start taking data. But it’s not ready yet, so for now we should fix this.
